[20기 최준식] MISSION 3: 로그인 엔드포인트 구현

JunsikChoi opened the pull request:
## Reference Branch: feature/junsik-signup

## 수정 사항 간략한 한줄 요약
- 로그인 엔드포인트 구현
		 
## 수정 사항들 자세한 내용
- urls.py에 SignInView 추가
- views.py에 SIgnInView Class 추가
- 예외 처리
    - JsonDecodeError : 데이터 형식 JSON 디코딩 실패 시 400 반환
    - KeyError: 필수 필드 (이메일, 비밀번호) 없을 시 400 반환
    - User DoesNotExist: DB에 User 존재하지 않을 시 401 반환
    - Password DoesNotMatch(AuthenticationError): 패스워드 매칭 안될 시 401 반환
    - JWT 관련 예외 처리
- 유저 조회 성공 시 200 반환
- JWT 로그인 기능 및 로그인 성공시 JWT 발행


## 기타 질문 및 특이 사항
없음

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
조건을 만족한다면 각 빈칸에 x 표시 해주세요
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.

JunsikChoi committed:
ADD: Basic SignIn Logic

- Add SignInView in urls.py
- Add SIgnInView Class in views.py
- Handling Exceptions for JsonDecodeError, KeyError, User DoesNotExist,
Password DoesNotMatch(AuthenticationError)
- Add Custom AuthenticationError to utils.py

JunsikChoi commented:
해당 브랜치 feature/junsik-signup에서 분기했기 때문에, 리뷰해주시면 feature/junsik-signin으로 먼저 merge 한 후 메인에 merge 시키겠습니다..!

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -64,3 +64,39 @@ def post(self, request):
+            return JsonResponse({"message": "INVALID_USER"}, status=401)
+
+        else:
+            return JsonResponse({"message": "SUCCESS"}, status=200)
```
except가 발생하지 않았다는 것은 로직이 성공적으로 돌았다는 이야기가 되겠죠? 그러면 성공 return else에 안 묶이고 except 문들 위 쪽으로 가도 되지 않을까요?

JunsikChoi replied:
수정완료!

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -64,3 +64,39 @@ def post(self, request):
+            user = User.objects.get(email=email)
+
+            # check if password matches
+            if user.password != str(password):
+                raise AuthenticationError
```
if 문에서 raise을 하는것 보다 바로 return 하는 것이 rasie을 위해 진행 되어야 하는 여러 단계를 안 거치고 바로 예외 처리 할 수 있지 않을까요? 여러 app에서 공통으로 사용되는 error 같은 겨우 exception을 커스터마이징 해서 사용하면 너무 좋겠지만, 지금 같은 경우는 조금 불필요해 보입니다.

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -64,3 +64,39 @@ def post(self, request):
             )
+
+class SignInView(View):
+    # Limit Http Method to POST
+    http_method_names = ["post"]
```
이 부분은 회원가입 때 남긴 코멘트 화인해 주세요.

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -64,3 +64,39 @@ def post(self, request):
+    def post(self, request):
+        
+        try:
+            # loads data from request body
```
각 기능마다 주석을 넣어 주어 어떤 기능을 나타내는지 잘 보여주었습니다. 주석 같은 경우 사람마다 스타일이 다르기 때문에 보통 회사에서는 통일된 스타일로 가는 경우가 많습니다. 위코드에서는 조금 더 깔끔하고, 주석에 대한 오해를 방지하고자 깃허브에 올라오는 코드에는 주석을 제거하고 있습니다. 주석 부분 제거 부탁드리겠습니다.

Kang-kyunghun reviewed (commented):
회원가입 때보다 확실히 깔끔해지고 가독성이 높아졌습니다!!!!!

JunsikChoi commented:
후후 심혈을 기울였습니다요..

JunsikChoi committed:
MODIFY: Apply Review Comments

JunsikChoi commented:
반영 완료했습니다!

JunsikChoi commented:
JSON response 부분을 해당 요청에 대한 짧은 코드를 보여주는 result 부분과 만약 가능하다면 자세한 내용을 보여주는 message 부분으로 나눠서 포맷 통일했어요.

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -41,4 +41,30 @@ def post(self, request):
+                return JsonResponse({"result": "INVALID_USER"}, status=401)
+
+            return JsonResponse({"result": "SUCCESS"}, status=200)
+
+        except JSONDecodeError as e:
+            return JsonResponse({"result": "JSON_DECODE_ERROR", "message": e.msg}, status=400)
```
실제로 에러와 에러 원인을 같이 반환하는 경우가 많기는 합니다. 그런데 그 때 보통 준식님이 result라고 한 부분을 message로 가고 message 부분을 key 또는 error 등 으로 하는 경우가 많습니다. 메시지 같은 경우는 프론트를 위해 무조건 해야 되는 부분이지만, 그 뒤에 부분은 회사마다 하는 경우도 다르고 형식도 다른 것 같습니다. 매번 느끼지만, 생각을 많이 하고 코딩 하는게 느껴지네요.

JunsikChoi replied:
감사합니다! 
그럼 에러 코드 부분을 message로 하고 지금 상세 에러 설명하는 message 부분은 더 좋은 네이밍 생각해보고 수정해보겠습니다!

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -41,4 +41,30 @@ def post(self, request):
+        try:
+            data = json.loads(request.body)
+
+            email = data['email']
+            password = data['password']
+
+            user = User.objects.get(email=email)
+
+            if user.password != str(password):
+                return JsonResponse({"result": "INVALID_USER"}, status=401)
+
+            return JsonResponse({"result": "SUCCESS"}, status=200)
```
암호화와 token 발급까지 한 번 추가해 보면 좋을 것 같아요

JunsikChoi replied:
네 bcrypt 이용해서 jwt 발급과 jwt 통한 로그인 구현해보겠습니다~

JunsikChoi committed:
MODIFY: models.py for longer password field

JunsikChoi committed:
MODIFY: Hashed Password Validation in SignInView

JunsikChoi committed:
MODIFY: SignIn with JWT

- SignIn with JWT when 'Authorization' Header exists
- Issue new JWT with new expire date when SignIn Success
- Handling JWT related exceptions

JunsikChoi committed:
FIX: Encode Error when Password is Number(int)

JunsikChoi committed:
MODIFY: Json Response Format

JunsikChoi committed:
FIX: Code Formatting

JunsikChoi committed:
RESOLVE CONFLICT

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -23,25 +27,70 @@ def post(self, request):
+    def post(self, request):
+        
+        try:
+            if request.headers.get('Authorization'):
+                token = request.headers.get('Authorization')
+                payload = jwt.decode(token, JWT_SECRET_KEY, JWT_ALGORITHM)
+                user = User.objects.get(pk=payload.get('user_id'))
```
로그인 기능에는 token을 디코딩하는 기능이 없어야 합니다. 왜냐하면 로그인을 해야 token이 발급 되기 때문입니다.

JunsikChoi replied:
헤더에 Authorization 토큰이 있는 경우 프론트에서 토큰 유효성을 확인한 후 확인되었을 경우 새로운 토큰을 발급받아 자동 로그인하는 경우를 구현한건데, 이 방법이 아니라면 자동 로그인은 어떤 방식으로 구현해야 할까요? :)

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -23,25 +27,70 @@ def post(self, request):
+                if not bcrypt.checkpw(str(data.get('password')).encode('utf-8'),user.password.encode('utf-8')):
+                    return JsonResponse({"status": "INVALID_USER"}, status=401)
+
+            new_token = jwt.encode({'user_id': user.id, 'iat': int(time.time()), 'exp': int(time.time()) + JWT_DURATION_SEC}, 
+                                JWT_SECRET_KEY, 
+                                JWT_ALGORITHM)
```
토큰의 만료시간에 iat까지 잘 적용하셨습니다. 실제로 저도 exp는 거의 적용해도 iat까지 적용 된 사례를 많이 못 봤었는데, 이번에도 많이 찾아 보셨네요~

JunsikChoi replied:
네 찾아보니까 jwt 디코딩 과정에서 exp가 있으면 자동으로 시간 체크해서 토큰 만료 여부까지 exception 발생 처리해주더라구요!

JunsikChoi committed:
MODIFY: Remove Signin with Token

JunsikChoi committed:
Merge branch 'feature/junsik-signin' of https://github.com/wecode-bootcamp-korea/20-Westagram-backend into feature/junsik-signin

JunsikChoi commented:
로그인시 토큰 디코딩 기능 제거했습니다 :)
